File: ahp/recorder.py
"""AHPRecorder -- the main SDK entry point.

Wires together: chain writer + evidence store + PII filters + signing +
witness client + context propagation into one automated flow.

Usage:
    from ahp.recorder import AHPRecorder

    recorder = AHPRecorder(agent_name="my-agent")
    # or with config file:
    recorder = AHPRecorder.from_config("ahp.yaml", agent_name="my-agent")

    # Record actions (fail-open -- never crashes the agent)
    recorder.record_action(
        tool_name="search_docs",
        parameters=b'{"query": "return policy"}',
        result=b'{"matches": [...]}',
        protocol=Protocol.MCP,
        action_type=ActionType.TOOL_CALL,
    )
"""

from __future__ import annotations

import json
import logging
import os
import tempfile
import threading
import time
from pathlib import Path
from typing import Callable, List, Optional

from ahp._base_recorder import RecorderBase
from ahp.config import AHPConfig, load_config
from ahp.core.chain import ChainWriter
from ahp.core.filters import Filter
from ahp.core.records import (
    ActionPayload,
    Authorization,
    Record,
    WitnessPayload,
)
from ahp.core.recovery import recover_chain
from ahp.core.signing import sign
from ahp.core.types import (
    ZERO_UUID,
    ActionType,
    AuthorizationType,
    GapReason,
    Protocol,
    ResultStatus,
)
from ahp.core.witness_client import send_checkpoint as _send_checkpoint

logger = logging.getLogger("ahp.recorder")

DEFAULT_MAX_SEGMENT_BYTES = 64 * 1024 * 1024  # 64MB


class AHPRecorder(RecorderBase):
    """Main SDK entry point -- wires all AHP components together.

    Fail-open by design: recording failures NEVER propagate to the host agent.
    All public methods that write records catch exceptions internally when used
    via :meth:`safe_record`.

    Lock hierarchy (acquire in this order to prevent deadlocks):
      1. self._recorder_lock (RLock) — protects counters and checkpoint logic
      2. self._chain._lock (Lock) — protects chain file I/O
    Never acquire _recorder_lock while holding _chain._lock.
    """

    # ------------------------------------------------------------------
    # Construction
    # ------------------------------------------------------------------

    def __init__(
        self,
        agent_name: str,
        chain_path: Optional[str] = None,
        level: int = 1,
        config: Optional[AHPConfig] = None,
        evidence_path: Optional[str] = None,
        checkpoint_interval: int = 1000,
        witness_interval: int = 1000,
        witness_endpoints: Optional[List[str]] = None,
        filter_presets: Optional[List[str]] = None,
        custom_filters: Optional[List[Filter]] = None,
        agent_framework: str = "",
        interceptors: Optional[List[str]] = None,
        on_record_written: Optional[Callable[[Record], None]] = None,
        on_error: Optional[Callable[[Exception, str], None]] = None,
    ) -> None:
        # ---- shared initialization (config, filters, evidence, signing) ----
        self._init_shared_components(
            agent_name=agent_name,
            level=level,
            config=config,
            evidence_path=evidence_path,
            filter_presets=filter_presets,
            custom_filters=custom_filters,
            agent_framework=agent_framework,
            interceptors=interceptors,
            checkpoint_interval=checkpoint_interval,
            witness_interval=witness_interval,
            witness_endpoints=witness_endpoints,
            on_record_written=on_record_written,
            on_error=on_error,
        )

        # ---- chain writer (with recovery + rotation) -----------------------
        if chain_path is None:
            chain_path = str(Path(tempfile.gettempdir()) / ("ahp_" + agent_name + ".ahp"))
        self._chain_path = chain_path

        # Recovery: if chain file already exists, scan and truncate corrupt tail
        self._recovery_result = None
        if Path(chain_path).exists():
            try:
                self._recovery_result = recover_chain(chain_path)
                if self._recovery_result.records_truncated > 0:
                    self._log_warning(
                        "Recovered chain %s: %d verified, %d truncated",
                        chain_path,
                        self._recovery_result.records_verified,
                        self._recovery_result.records_truncated,
                    )
            except Exception as exc:
                self._log_warning("Chain recovery failed for %s", chain_path, exc_info=True)
                self._fire_error_callback(exc, "chain_recovery")

        # Continue chain from recovery state if available
        if self._recovery_result is not None and self._recovery_result.records_verified > 0:
            self._chain = ChainWriter(
                chain_path,
                prev_hash=self._recovery_result.last_prev_hash,
                start_sequence=self._recovery_result.last_valid_seq,
            )
        else:
            self._chain = ChainWriter(chain_path)

        # Rotation support: track segment size limit
        self._max_segment_bytes = DEFAULT_MAX_SEGMENT_BYTES  # 64MB

        # Update evidence path if not explicitly set
        if self._evidence_enabled and self._evidence is not None:
            if evidence_path is None:
                epath = str(Path(chain_path).parent / "evidence")
                from ahp.core.evidence import EvidenceStore

                self._evidence = EvidenceStore(epath)

        # ---- concurrency lock for counters -----------------------------------
        self._recorder_lock = threading.RLock()

        # ---- emit genesis records ------------------------------------------
        self._emit_boot_record()
        if self._level >= 2 and self._keypair is not None:
            self._emit_key_genesis_record()

        # ---- emit recovery + gap records if recovery found corrupt data ----
        if self._recovery_result is not None and self._recovery_result.records_truncated > 0:
            self._emit_recovery_records(self._recovery_result)

    # ------------------------------------------------------------------
    # Class-method constructors
    # ------------------------------------------------------------------

    @classmethod
    def from_config(
        cls,
        config_path: str,
        agent_name: str,
        chain_path: Optional[str] = None,
        evidence_path: Optional[str] = None,
    ) -> "AHPRecorder":
        """Create an AHPRecorder from a YAML/JSON configuration file."""
        cfg = load_config(config_path, agent_name=agent_name)
        return cls(
            agent_name=agent_name,
            config=cfg,
            chain_path=chain_path,
            evidence_path=evidence_path,
        )

    # ------------------------------------------------------------------
    # Core recording methods
    # ------------------------------------------------------------------

    def record_action(
        self,
        tool_name: str,
        parameters: bytes = b"",
        result: bytes = b"",
        protocol: Protocol = Protocol.CUSTOM,
        action_type: ActionType = ActionType.TOOL_CALL,
        result_status: ResultStatus = ResultStatus.SUCCESS,
        response_time_ms: int = 0,
        target_entity: str = "",
        parent_action_id: bytes = ZERO_UUID,
        authorization: Optional[Authorization] = None,
        model_id: str = "",
        input_token_count: int = 0,
        output_token_count: int = 0,
    ) -> Record:
        """Record a single agent action.

        This is the primary recording method.  It filters PII, hashes
        content, optionally stores evidence, writes to the chain, and
        triggers checkpoints / witness submissions when their intervals
        are reached.

        Returns the :class:`Record` written to the chain.
        """
        # Phase 1: Outside lock — pure computation + idempotent I/O
        param_hash, result_hash, filtered_params, filtered_result, redacted = self._filter_action_payloads(
            parameters, result
        )
        evidence_uri = self._store_evidence(filtered_params, filtered_result, param_hash)

        # Apply PII filters to string fields that go directly into the chain
        if target_entity and self._filters.filters:
            filtered_te, te_redacted = self._filters.apply(target_entity.encode("utf-8"), scope="parameters")
            target_entity = filtered_te.decode("utf-8")
            if te_redacted:
                redacted = True

        # Phase 2: Inside lock — state mutations only
        with self._recorder_lock:
            self._flush_pending_gap()

            payload = ActionPayload(
                parent_action_id=parent_action_id,
                tool_name=tool_name,
                parameters_hash=param_hash,
                result_hash=result_hash,
                result_status=result_status,
                response_time_ms=response_time_ms,
                protocol=protocol,
                action_type=action_type,
                target_entity=target_entity,
                evidence_uri=evidence_uri,
                redacted=redacted,
                model_id=model_id,
                input_token_count=input_token_count,
                output_token_count=output_token_count,
                authorization=authorization or Authorization(type=AuthorizationType.AUTH_NONE),
            )

            record = self._chain.write_record(payload)
            self._track_record(record)

            if self._records_since_checkpoint >= self._checkpoint_interval:
                self.emit_checkpoint()

            if self._level >= 3 and self._witness_endpoints and self._records_since_witness >= self._witness_interval:
                self.send_witness_checkpoint()

            self._check_rotation()

            return record

    def record_inference(
        self,
        tool_name: str,
        parameters: bytes = b"",
        result: bytes = b"",
        model_id: str = "",
        input_token_count: int = 0,
        output_token_count: int = 0,
        protocol: Protocol = Protocol.CUSTOM,
        result_status: ResultStatus = ResultStatus.SUCCESS,
        response_time_ms: int = 0,
        target_entity: str = "",
        parent_action_id: bytes = ZERO_UUID,
        authorization: Optional[Authorization] = None,
    ) -> Record:
        """Record an inference (LLM) call.

        Convenience wrapper around :meth:`record_action` that sets
        ``action_type=INFERENCE`` and populates model / token fields.
        """
        return self.record_action(
            tool_name=tool_name,
            parameters=parameters,
            result=result,
            protocol=protocol,
            action_type=ActionType.INFERENCE,
            result_status=result_status,
            response_time_ms=response_time_ms,
            target_entity=target_entity,
            parent_action_id=parent_action_id,
            authorization=authorization,
            model_id=model_id,
            input_token_count=input_token_count,
            output_token_count=output_token_count,
        )

    # ------------------------------------------------------------------
    # Checkpointing
    # ------------------------------------------------------------------

    def emit_checkpoint(self) -> Record:
        """Emit a checkpoint record.

        Computes the Merkle root of records since the last checkpoint,
        signs it when ``level >= 2``, and writes a
        :class:`CheckpointPayload` to the chain.
        """
        payload = self._build_checkpoint_payload(
            record_count=self._chain.record_count + 1,
            gap_count=self._chain.gap_count,
            chain_hash=self._chain.prev_hash,
        )

        record = self._chain.write_record(payload)

        # Reset counters (shared logic from base)
        self._reset_checkpoint_counters()

        return record

    # ------------------------------------------------------------------
    # Witness
    # ------------------------------------------------------------------

    def send_witness_checkpoint(self) -> None:
        """Send the current chain state to all configured witness endpoints.

        On receipt, a :class:`WitnessPayload` is written to the chain.
        On failure, the error is logged but does NOT block the agent
        (per spec Section 8.5).
        """
        agent_id_hex = self._chain.agent_id.hex()
        chain_hash_hex = self._chain.prev_hash.hex()
        sequence = self._chain.sequence
        timestamp_ms = int(time.time() * 1000)

        sig_hex = ""
        key_id_hex = ""
        public_key_hex = ""
        if self._level >= 2 and self._keypair is not None:
            # Sign canonical JSON of checkpoint fields (must match witness verification)
            sign_data = json.dumps(
                {
                    "agent_id": agent_id_hex,
                    "chain_hash": chain_hash_hex,
                    "sequence": sequence,
                    "timestamp_ms": timestamp_ms,
                },
                sort_keys=True,
            ).encode()
            sig_hex = sign(sign_data, self._keypair.private_key_bytes).hex()
            key_id_hex = self._keypair.key_id.hex()
            public_key_hex = self._keypair.public_key_bytes.hex()

        for endpoint in self._witness_endpoints:
            try:
                receipt = _send_checkpoint(
                    endpoint=endpoint,
                    agent_id=agent_id_hex,
                    chain_hash=chain_hash_hex,
                    sequence=sequence,
                    timestamp_ms=timestamp_ms,
                    signature=sig_hex,
                    signing_key_id=key_id_hex,
                    public_key=public_key_hex,
                )
                if receipt is not None:
                    witness_payload = WitnessPayload(
                        witness_id=receipt.get("witness_id", endpoint),
                        checkpoint_seq=sequence,
                        checkpoint_hash=self._chain.prev_hash,
                        witness_timestamp=receipt.get("timestamp_ms", timestamp_ms),
                        receipt_signature=bytes.fromhex(receipt.get("signature", "00" * 64)),
                        witness_public_key=bytes.fromhex(receipt.get("public_key", "00" * 32)),
                    )
                    self._chain.write_record(witness_payload)
            except Exception as exc:
                self._log_warning("Witness checkpoint to %s failed", endpoint, exc_info=True)
                self._fire_error_callback(exc, "witness_checkpoint")

        self._reset_witness_counter()

    # ------------------------------------------------------------------
    # Fail-open wrapper
    # ------------------------------------------------------------------

    def safe_record(self, **kwargs) -> Optional[Record]:  # type: ignore[return]
        """Fail-open wrapper around :meth:`record_action`.

        If recording raises an exception the error is captured and a
        :class:`GapRecord` will be emitted on the next successful write.
        The host agent is never disrupted.
        """
        try:
            return self.record_action(**kwargs)
        except Exception as exc:
            # Remember the gap so the next successful write can document it.
            if not self._pending_gap:
                self._gap_first_lost_seq = self._chain.sequence + 1
            self._pending_gap = True
            self._gap_reason = "INTERCEPTOR_FAILURE"
            self._gap_detail = str(exc)
            self._log_warning("AHP safe_record failed (will emit GapRecord): %s", exc)
            self._fire_error_callback(exc, "safe_record")
            return None

    # ------------------------------------------------------------------
    # Resource management
    # ------------------------------------------------------------------

    def close(self) -> None:
        """Release resources held by the recorder (chain writer, file locks)."""
        self._chain.close()

    def __enter__(self) -> "AHPRecorder":
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.close()

    # ------------------------------------------------------------------
    # Read-only accessors
    # ------------------------------------------------------------------

    @property
    def chain_path(self) -> str:
        """Path to the chain file."""
        return str(self._chain.path)

    @property
    def chain(self) -> ChainWriter:
        """Underlying :class:`ChainWriter` (read-only access)."""
        return self._chain

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _emit_boot_record(self) -> None:
        """Write the BootRecord as the very first record in the chain."""
        payload = self._build_boot_payload()
        record = self._chain.write_record(payload)
        self._track_record(record)

    def _emit_key_genesis_record(self) -> None:
        """Write a KeyGenesisRecord (level >= 2)."""
        payload = self._build_key_genesis_payload()
        if payload is None:
            return
        record = self._chain.write_record(payload)
        self._track_record(record)

    def _flush_pending_gap(self) -> None:
        """If a previous safe_record failed, emit a GapRecord now.

        Called from record_action() which already holds _recorder_lock.
        """
        if not self._pending_gap:
            return

        first_lost = self._gap_first_lost_seq
        last_lost = self._chain.sequence

        if first_lost > last_lost:
            last_lost = first_lost

        reason = GapReason.INTERCEPTOR_FAILURE
        gap_record = self._chain.write_gap(
            first_lost=first_lost,
            last_lost=last_lost,
            reason=reason,
            detail=self._gap_detail,
        )
        self._track_record(gap_record)

        self._pending_gap = False
        self._gap_reason = ""
        self._gap_detail = ""
        self._gap_first_lost_seq = 0

    def _emit_recovery_records(self, recovery_result) -> None:
        """Emit RecoveryRecord + GapRecord after crash recovery (per spec Section 3.6)."""
        from ahp.core.types import RecoveryMethod

        recovery_record = self._chain.write_recovery(
            records_verified=recovery_result.records_verified,
            records_truncated=recovery_result.records_truncated,
            last_valid_seq=recovery_result.last_valid_seq,
            method=RecoveryMethod.CHAIN_SCAN,
            detail="Automatic crash recovery on startup",
        )
        self._track_record(recovery_record)

        if recovery_result.records_truncated > 0:
            first_lost = recovery_result.last_valid_seq + 1
            last_lost = first_lost + recovery_result.records_truncated - 1
            gap_record = self._chain.write_gap(
                first_lost=first_lost,
                last_lost=last_lost,
                reason=GapReason.CRASH,
                detail="Records lost during crash recovery",
            )
            self._track_record(gap_record)

    def _check_rotation(self) -> None:
        """Rotate the chain file if it exceeds 64MB."""
        chain_size = self._chain.bytes_written

        if chain_size < self._max_segment_bytes:
            return

        logger.info(
            "Chain file %s reached %d bytes, rotating",
            self._chain_path,
            chain_size,
        )

        prev_hash = self._chain.prev_hash
        prev_sequence = self._chain.sequence

        self._chain.close()

        timestamp = int(time.time())
        segment_path = self._chain_path + f".{timestamp}.segment"
        try:
            os.rename(self._chain_path, segment_path)
        except OSError:
            self._log_warning("Failed to rename chain for rotation", exc_info=True)

        self._chain = ChainWriter(
            self._chain_path,
            prev_hash=prev_hash,
            start_sequence=prev_sequence,
        )

        self._emit_boot_record()
        if self._level >= 2 and self._keypair is not None:
            self._emit_key_genesis_record()

File: packages/sdk-typescript/src/recorder.ts
/**
 * AHPRecorder — the main SDK entry point.
 *
 * Wires together: chain writer + evidence store + PII filters + signing +
 * crash recovery into one automated flow.
 *
 * Usage:
 *   import { AHPRecorder } from "@ahp/sdk";
 *
 *   const recorder = new AHPRecorder({ agentName: "my-agent" });
 *
 *   recorder.recordAction({
 *     toolName: "search_docs",
 *     parameters: Buffer.from('{"query": "return policy"}'),
 *     result: Buffer.from('{"matches": [...]}'),
 *     protocol: Protocol.MCP,
 *     actionType: ActionType.TOOL_CALL,
 *   });
 *
 *   recorder.close();
 */

import * as os from "os";
import * as path from "path";
import * as crypto from "crypto";
import * as fs from "fs";

import {
  Record as AHPRecord,
  ResultStatus,
  Protocol,
  ActionType,
  AuthorizationType,
  GapReason,
  ChainLevel,
  FsyncMode,
  ZERO_HASH_32,
  ZERO_UUID,
  createActionPayload,
  createBootPayload,
  createCheckpointPayload,
  createKeyPayload,
  Authorization,
} from "./types";
import { canonicalBytes } from "./canonical";
import { ChainWriter } from "./chain";
import { EvidenceStore } from "./evidence";
import { Filter, FilterPipeline } from "./filters";
import {
  KeyPair,
  generateKeypair,
  sign as ed25519Sign,
  computeMerkleRoot,
} from "./signing";
import { AHPConfig, loadConfig, defaultConfig } from "./config";
import { recoverChain, RecoveryResult } from "./recovery";

// SDK identity constants
const SDK_NAME = "ahp-typescript";
const SDK_VERSION = "0.1.0";

// Map string fsync modes from config to enum values
const FSYNC_MAP: { [key: string]: FsyncMode } = {
  every: FsyncMode.EVERY,
  batch: FsyncMode.BATCH,
  none: FsyncMode.NONE,
};

// Default max segment size: 64MB
const DEFAULT_MAX_SEGMENT_BYTES = 64 * 1024 * 1024;

export interface AHPRecorderOptions {
  agentName: string;
  chainPath?: string;
  level?: number;
  config?: AHPConfig;
  evidencePath?: string;
  checkpointInterval?: number;
  witnessInterval?: number;
  witnessEndpoints?: string[];
  filterPresets?: string[];
  customFilters?: Filter[];
  agentFramework?: string;
  interceptors?: string[];
}

export interface RecordActionOptions {
  toolName: string;
  parameters?: Uint8Array;
  result?: Uint8Array;
  protocol?: Protocol;
  actionType?: ActionType;
  resultStatus?: ResultStatus;
  responseTimeMs?: number;
  targetEntity?: string;
  parentActionId?: Uint8Array;
  authorization?: Authorization;
  modelId?: string;
  inputTokenCount?: number;
  outputTokenCount?: number;
}

export class AHPRecorder {
  private _cfg: AHPConfig;
  private _agentName: string;
  private _level: number;
  private _checkpointInterval: number;
  private _agentFramework: string;
  private _interceptors: string[];

  private _chain: ChainWriter;
  private _chainPath: string;
  private _maxSegmentBytes: number;

  private _evidenceEnabled: boolean;
  private _evidence: EvidenceStore | null;

  private _filters: FilterPipeline;
  private _keypair: KeyPair | null;

  // Internal counters
  private _recordsSinceCheckpoint: number = 0;
  private _recordHashesSinceCheckpoint: Uint8Array[] = [];

  // Fail-open gap state
  private _pendingGap: boolean = false;
  private _gapDetail: string = "";
  private _gapFirstLostSeq: bigint = 0n;

  private _recoveryResult: RecoveryResult | null = null;

  constructor(options: AHPRecorderOptions) {
    const {
      agentName,
      chainPath,
      level = 1,
      config,
      evidencePath,
      checkpointInterval = 1000,
      filterPresets,
      customFilters,
      agentFramework = "",
      interceptors = [],
    } = options;

    // Resolve config
    if (config) {
      this._cfg = config;
    } else {
      this._cfg = defaultConfig();
      this._cfg.level = level;
      this._cfg.agentName = agentName;
      this._cfg.agentFramework = agentFramework;
      this._cfg.checkpointInterval = checkpointInterval;
      if (filterPresets) {
        this._cfg.filterPresets = filterPresets;
      }
    }

    this._agentName = agentName;
    this._level = this._cfg.level;
    this._checkpointInterval = this._cfg.checkpointInterval;
    this._agentFramework = this._cfg.agentFramework || agentFramework;
    this._interceptors = interceptors;
    this._maxSegmentBytes = DEFAULT_MAX_SEGMENT_BYTES;

    // Chain writer (with recovery)
    this._chainPath =
      chainPath ||
      path.join(os.tmpdir(), `ahp_${agentName}.ahp`);

    // Recovery: if chain file already exists, scan and truncate corrupt tail
    if (fs.existsSync(this._chainPath)) {
      try {
        this._recoveryResult = recoverChain(this._chainPath);
      } catch {
        // Recovery failed; proceed with fresh chain
      }
    }

    this._chain = new ChainWriter(this._chainPath);

    // Evidence store
    this._evidenceEnabled = this._cfg.evidenceRecord;
    this._evidence = null;
    if (this._evidenceEnabled) {
      const epath =
        evidencePath ||
        path.join(path.dirname(this._chainPath), "evidence");
      this._evidence = new EvidenceStore(epath);
    }

    // PII filter pipeline
    const presetList = [...this._cfg.filterPresets];
    if (filterPresets) {
      for (const p of filterPresets) {
        if (!presetList.includes(p)) {
          presetList.push(p);
        }
      }
    }

    const customFilterList: Filter[] = [...(customFilters || [])];
    for (const fc of this._cfg.filters) {
      customFilterList.push(
        new Filter(fc.name, fc.pattern, fc.replacement, [...fc.scope])
      );
    }

    this._filters = new FilterPipeline(
      customFilterList.length > 0 ? customFilterList : null,
      presetList.length > 0 ? presetList : null
    );

    // Signing (level >= 2)
    this._keypair = null;
    if (this._level >= 2) {
      this._keypair = generateKeypair();
    }

    // Emit genesis records
    this._emitBootRecord();
    if (this._level >= 2 && this._keypair !== null) {
      this._emitKeyGenesisRecord();
    }

    // Emit recovery + gap records if recovery found corrupt data
    if (
      this._recoveryResult !== null &&
      this._recoveryResult.recordsTruncated > 0
    ) {
      this._emitRecoveryRecords(this._recoveryResult);
    }
  }

  /**
   * Create an AHPRecorder from a YAML/JSON configuration file.
   */
  static fromConfig(
    configPath: string,
    agentName: string,
    chainPath?: string,
    evidencePath?: string
  ): AHPRecorder {
    const cfg = loadConfig(configPath, agentName);
    return new AHPRecorder({
      agentName,
      config: cfg,
      chainPath,
      evidencePath,
    });
  }

  // --- Core recording methods ---

  /**
   * Record a single agent action.
   *
   * Filters PII, hashes content, optionally stores evidence,
   * writes to the chain, and triggers checkpoints when interval reached.
   */
  recordAction(opts: RecordActionOptions): AHPRecord {
    const {
      toolName,
      parameters = new Uint8Array(0),
      result = new Uint8Array(0),
      protocol = Protocol.CUSTOM,
      actionType = ActionType.TOOL_CALL,
      resultStatus = ResultStatus.SUCCESS,
      responseTimeMs = 0,
      targetEntity = "",
      parentActionId = new Uint8Array(ZERO_UUID),
      authorization,
      modelId = "",
      inputTokenCount = 0,
      outputTokenCount = 0,
    } = opts;

    // 0. Flush pending gap from previous failure
    this._flushPendingGap();

    // 1. Apply PII filters
    const [paramHash, filteredParams, paramRedacted] =
      this._filters.hashPayload(parameters, "parameters");
    const [resultHash, filteredResult, resultRedacted] =
      this._filters.hashPayload(result, "results");
    const redacted = paramRedacted || resultRedacted;

    // 2. Store evidence if configured
    let evidenceUri = "";
    if (
      this._evidence !== null &&
      (filteredParams.length > 0 || filteredResult.length > 0)
    ) {
      if (filteredParams.length > 0) {
        this._evidence.store(filteredParams);
      }
      if (filteredResult.length > 0) {
        this._evidence.store(filteredResult);
      }
      evidenceUri = "evidence://" + Buffer.from(paramHash).toString("hex");
    }

    // 3. Build payload
    const payload = createActionPayload({
      parent_action_id: parentActionId,
      tool_name: toolName,
      parameters_hash: paramHash,
      result_hash: resultHash,
      result_status: resultStatus,
      response_time_ms: responseTimeMs,
      protocol,
      action_type: actionType,
      target_entity: targetEntity,
      evidence_uri: evidenceUri,
      redacted,
      model_id: modelId,
      input_token_count: inputTokenCount,
      output_token_count: outputTokenCount,
      authorization: authorization || {
        type: AuthorizationType.AUTH_NONE,
        entries: [],
      },
    });

    // 4. Write to chain
    const record = this._chain.writeRecord(payload);

    // 5. Track checkpoint state
    this._trackRecord(record);

    // 6. Auto-checkpoint
    if (this._recordsSinceCheckpoint >= this._checkpointInterval) {
      this.emitCheckpoint();
    }

    // 7. Auto-rotate if chain exceeds 64MB
    this._checkRotation();

    return record;
  }

  /**
   * Record an inference (LLM) call.
   *
   * Convenience wrapper around recordAction that sets actionType=INFERENCE.
   */
  recordInference(
    opts: Omit<RecordActionOptions, "actionType"> & { actionType?: ActionType }
  ): AHPRecord {
    return this.recordAction({
      ...opts,
      actionType: ActionType.INFERENCE,
    });
  }

  // --- Checkpointing ---

  /**
   * Emit a checkpoint record.
   *
   * Computes the Merkle root of records since the last checkpoint,
   * signs it when level >= 2, and writes a CheckpointPayload.
   */
  emitCheckpoint(): AHPRecord {
    const merkleRoot = computeMerkleRoot(
      this._recordHashesSinceCheckpoint
    );

    let signature: Uint8Array = new Uint8Array(64);
    let signingKeyId: Uint8Array = new Uint8Array(ZERO_HASH_32);
    if (this._level >= 2 && this._keypair !== null) {
      signature = new Uint8Array(ed25519Sign(merkleRoot, this._keypair.privateKeyBytes));
      signingKeyId = new Uint8Array(this._keypair.keyId);
    }

    const evidenceStatus = this._getEvidenceStatus();

    const payload = createCheckpointPayload({
      record_count: BigInt(this._chain.recordCount + 1),
      gap_count: BigInt(this._chain.gapCount),
      chain_hash: new Uint8Array(this._chain.prevHash),
      merkle_root: merkleRoot,
      signature,
      signing_key_id: signingKeyId,
      evidence_available: BigInt(evidenceStatus.available),
      evidence_exported: BigInt(evidenceStatus.exported),
      evidence_expired: BigInt(evidenceStatus.expired),
      evidence_missing: BigInt(evidenceStatus.missing),
    });

    const record = this._chain.writeRecord(payload);

    // Reset counters
    this._recordsSinceCheckpoint = 0;
    this._recordHashesSinceCheckpoint = [];

    return record;
  }

  // --- Fail-open wrapper ---

  /**
   * Fail-open wrapper around recordAction.
   *
   * If recording raises an exception the error is captured and a
   * GapRecord will be emitted on the next successful write.
   */
  safeRecord(opts: RecordActionOptions): AHPRecord | null {
    try {
      return this.recordAction(opts);
    } catch (e) {
      if (!this._pendingGap) {
        this._gapFirstLostSeq = this._chain.sequence + 1n;
      }
      this._pendingGap = true;
      this._gapDetail = String(e);
      return null;
    }
  }

  // --- Resource management ---

  close(): void {
    this._chain.close();
  }

  // --- Read-only accessors ---

  get chainPath(): string {
    return this._chainPath;
  }

  get chain(): ChainWriter {
    return this._chain;
  }

  get level(): number {
    return this._level;
  }

  get keypair(): KeyPair | null {
    return this._keypair;
  }

  get evidenceStore(): EvidenceStore | null {
    return this._evidence;
  }

  get filterPipeline(): FilterPipeline {
    return this._filters;
  }

  // --- Internal helpers ---

  private _emitBootRecord(): void {
    const runtimeInfo = `node ${process.version} / ${process.platform}`;
    const fsyncMode = FSYNC_MAP[this._cfg.fsyncMode] ?? FsyncMode.BATCH;

    const payload = createBootPayload({
      sdk_name: SDK_NAME,
      sdk_version: SDK_VERSION,
      interceptors: this._interceptors,
      agent_framework: this._agentFramework,
      agent_name: this._agentName,
      runtime: runtimeInfo,
      chain_level: this._level as ChainLevel,
      fsync_mode: fsyncMode,
      clock_source: "system",
      inference_recording: this._cfg.inferenceRecord,
      inference_evidence: this._cfg.inferenceEvidence,
      evidence_recording: this._cfg.evidenceRecord,
      filter_config_hash: this._filters.configHash(),
      matched_agent_rule: this._cfg.matchedAgentRule,
      config_source: this._cfg.configSource,
      authorization_recording: this._cfg.authorizationRecord,
    });

    const record = this._chain.writeRecord(payload);
    this._trackRecord(record);
  }

  private _emitKeyGenesisRecord(): void {
    if (this._keypair === null) return;

    const payload = createKeyPayload({
      public_key: this._keypair.publicKeyBytes,
      key_id: this._keypair.keyId,
      expires_at: 0n, // no expiry for session keys
      supersedes_key_id: new Uint8Array(ZERO_HASH_32),
    });

    const record = this._chain.writeRecord(payload);
    this._trackRecord(record);
  }

  private _trackRecord(record: AHPRecord): void {
    this._recordsSinceCheckpoint += 1;

    // Keep the SHA-256 of the canonical bytes for Merkle tree
    const stored = canonicalBytes(record);
    this._recordHashesSinceCheckpoint.push(
      new Uint8Array(crypto.createHash("sha256").update(stored).digest())
    );
  }

  private _flushPendingGap(): void {
    if (!this._pendingGap) return;

    const firstLost = this._gapFirstLostSeq;
    let lastLost = this._chain.sequence;

    if (firstLost > lastLost) {
      lastLost = firstLost;
    }

    const gapRecord = this._chain.writeGap(
      firstLost,
      lastLost,
      GapReason.INTERCEPTOR_FAILURE,
      this._gapDetail
    );
    this._trackRecord(gapRecord);

    this._pendingGap = false;
    this._gapDetail = "";
    this._gapFirstLostSeq = 0n;
  }

  private _emitRecoveryRecords(recoveryResult: RecoveryResult): void {
    const recoveryRecord = this._chain.writeRecovery(
      BigInt(recoveryResult.recordsVerified),
      BigInt(recoveryResult.recordsTruncated),
      recoveryResult.lastValidSeq
    );
    this._trackRecord(recoveryRecord);

    if (recoveryResult.recordsTruncated > 0) {
      const firstLost = recoveryResult.lastValidSeq + 1n;
      const lastLost =
        firstLost + BigInt(recoveryResult.recordsTruncated) - 1n;
      const gapRecord = this._chain.writeGap(
        firstLost,
        lastLost,
        GapReason.CRASH,
        "Records lost during crash recovery"
      );
      this._trackRecord(gapRecord);
    }
  }

  private _checkRotation(): void {
    const chainSize = this._chain.bytesWritten;

    if (chainSize < this._maxSegmentBytes) return;

    // Close persistent file handle before renaming
    this._chain.close();

    // Rename current chain to a timestamped segment
    const timestamp = Math.floor(Date.now() / 1000);
    const segmentPath = this._chainPath + `.${timestamp}.segment`;
    try {
      fs.renameSync(this._chainPath, segmentPath);
    } catch {
      return; // Failed to rename; skip rotation
    }

    // Open a fresh chain writer — note: ChainWriter constructor creates a new file
    this._chain = new ChainWriter(this._chainPath);

    // Emit genesis records in the fresh segment
    this._emitBootRecord();
    if (this._level >= 2 && this._keypair !== null) {
      this._emitKeyGenesisRecord();
    }
  }

  private _getEvidenceStatus(): {
    available: number;
    exported: number;
    expired: number;
    missing: number;
  } {
    if (this._evidence !== null) {
      const counts = this._evidence.count();
      return {
        available: counts.available,
        exported: 0,
        expired: 0,
        missing: counts.missing,
      };
    }
    return { available: 0, exported: 0, expired: 0, missing: 0 };
  }
}
